### PR TITLE
Optimism morpho token

### PR DESCRIPTION
**Pull request by Jean-Grimal**

### CODE REVIEW

File: src/OptimismMintableERC20Upgradeable.sol [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     * @return Whether or not the interface is supported by this contract.
+     */
+    function supportsInterface(bytes4 _interfaceId) external pure returns (bool) {
+        bytes4 iface1 = type(IERC165).interfaceId;
+        // Interface corresponding to the legacy L2StandardERC20.
+        bytes4 iface2 = type(ILegacyMintableERC20).interfaceId;
+        // Interface corresponding to the updated OptimismMintableERC20 (this contract).
+        bytes4 iface3 = type(IOptimismMintableERC20).interfaceId;
+        return _interfaceId == iface1 || _interfaceId == iface2 || _interfaceId == iface3;
+    }
```
Comment (Jean-Grimal) [inline src/OptimismMintableERC20Upgradeable.sol:96]: Don't know if we want to keep this

Comment (MerlinEgalite) [reply src/OptimismMintableERC20Upgradeable.sol:96]: could be removed i think

Comment (MerlinEgalite) [reply src/OptimismMintableERC20Upgradeable.sol:96]: can you check the bridge contract?

File: src/ERC20DelegatesUpgradeable.sol [hunk 2]
```diff
@@ -12,13 +12,13 @@ import {EIP712Upgradeable} from
 import {Initializable} from "lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol";
 
 /// @title ERC20DelegatesUpgradeable
-/// @author Morpho Labs
+/// @author Morpho Association
```
Comment (QGarchery) [inline src/ERC20DelegatesUpgradeable.sol:15]: Why this change ? The author is Morpho Labs, but it is written for the association, right ?

File: src/OptimismMintableERC20Upgradeable.sol [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    /// @dev Sets the values for {remoteToken} and {bridge}.
+    /// @dev All two of these values are immutable: they can only be set once during initialization.
+    function __OptimismMintableERC20_init(address remoteToken_, address bridge_) internal onlyInitializing {
+        __OptimismMintableERC20_init_unchained(remoteToken_, bridge_);
+    }
+
+    function __OptimismMintableERC20_init_unchained(address remoteToken_, address bridge_) internal onlyInitializing {
+        require(remoteToken_ != address(0), ZeroAddress());
+        require(bridge_ != address(0), ZeroAddress());
```
Comment (QGarchery) [inline src/OptimismMintableERC20Upgradeable.sol:67]: Those checks are done twice: once here, and once in the MorphoTokenOptimism `initialize` function

File: src/interfaces/IOptimismMintableERC20.sol
Comment (QGarchery) [inline src/interfaces/IOptimismMintableERC20.sol:1]: Could we import it instead, from [here](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/IOptimismMintableERC20.sol) ?

File: test/MorphoTokenOptimism.t.sol [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        vm.assume(account != BRIDGE);
+        amount = bound(amount, MIN_TEST_AMOUNT, MAX_TEST_AMOUNT);
+
+        vm.expectRevert();
+        vm.prank(account);
+        morphoOptimism.mint(to, amount);
+    }
+
+    function testMint(address account, uint256 amount) public {
+        vm.assume(account != address(0));
```
Comment (QGarchery) [inline test/MorphoTokenOptimism.t.sol:84]:
```suggestion
        vm.assume(to != address(0));
```
otherwise a bit confusing with the test just above

File: test/MorphoTokenOptimism.t.sol [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        vm.assume(account != BRIDGE);
+        amount = bound(amount, MIN_TEST_AMOUNT, MAX_TEST_AMOUNT);
+
+        vm.expectRevert();
+        vm.prank(account);
+        morphoOptimism.burn(from, amount);
+    }
+
+    function testBurn(address account, uint256 amountMinted, uint256 amountBurned) public {
+        vm.assume(account != address(0));
```
Comment (QGarchery) [inline test/MorphoTokenOptimism.t.sol:109]:
```suggestion
        vm.assume(from != address(0));
```
same

File: test/MorphoTokenOptimism.t.sol [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        morphoOptimism = MorphoTokenOptimism(payable(address(tokenProxy)));
+        morphoOptimism.initialize(MORPHO_DAO, REMOTE_TOKEN, BRIDGE);
+    }
+
+    function testInitializeZeroAddress(address randomAddress) public {
+        vm.assume(randomAddress != address(0));
+
+        address proxy = address(new ERC1967Proxy(address(tokenImplem), hex""));
+
+        vm.expectRevert();
```
Comment (QGarchery) [inline test/MorphoTokenOptimism.t.sol:39]: We could specify the `expectRevert` in this file, to have the matching error

File: src/OptimismMintableERC20Upgradeable.sol
Comment (QGarchery) [inline src/OptimismMintableERC20Upgradeable.sol:1]: Should we (and is it possible to) confirm that this works as intended, and that this will make it possible to bridge tokens ? Potentially an integration test that mimicks the bridging, and that would catch if we forgot to add the burn functionality for example

File: src/OptimismMintableERC20Upgradeable.sol [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        emit Burn(from, amount);
+    }
+
+    /// @notice ERC165 interface check function.
+    /// @param _interfaceId Interface ID to check.
+    /// @return Whether or not the interface is supported by this contract.
+    function supportsInterface(bytes4 _interfaceId) external pure returns (bool) {
+        bytes4 iface1 = type(IERC165).interfaceId;
+        // Interface corresponding to the updated OptimismMintableERC20 (this contract).
+        bytes4 iface3 = type(IOptimismMintableERC20).interfaceId;
```
Comment (QGarchery) [inline src/OptimismMintableERC20Upgradeable.sol:94]:
```suggestion
        bytes4 interfaceERC165 = type(IERC165).interfaceId;
        bytes4 interfaceOptimismMintableERC20 = type(IOptimismMintableERC20).interfaceId;
```

File: test/MorphoToken.t.sol [hunk 8]
```diff
@@ -201,4 +201,10 @@ contract MorphoTokenTest is BaseTest {
         assertEq(newMorpho.getVotes(delegatee1), initialAmount - transferredAmount);
         assertEq(newMorpho.getVotes(delegatee2), transferredAmount);
     }
+
+    function testERC20DelegatesStorageLocation() public pure {
+        bytes32 expected =
+            keccak256(abi.encode(uint256(keccak256("morpho.storage.ERC20Delegates")) - 1)) & ~bytes32(uint256(0xff));
+        assertEq(expected, 0x1dc92b2c6e971ab6e08dfd7dcec0e9496d223ced663ba2a06543451548549500);
```
Comment (QGarchery) [inline test/MorphoToken.t.sol:208]: Maybe this would be a reason to have storage locations internal instead of private, so we can test them. For storage locations defined by the lib it's fine, because they should be tested independently, but for our added definitions it could be beneficial

File: src/OptimismMintableERC20Upgradeable.sol (hunk 1 above)
Comment (QGarchery) [reply src/OptimismMintableERC20Upgradeable.sol]: > Don't know if we want to keep this

`IERC165` is included in the interface `IOptimismMintableERC20` that is inherited from the contract, so maybe it's necessary to comply to ERC165

File: test/MorphoTokenOptimism.t.sol (hunk 6 above)
Comment (Jean-Grimal) [reply test/MorphoTokenOptimism.t.sol:39]: As the `require` takes a custom error (and not a string) as imput I couldn't manage to match the error in the `expectRevert`

I gave the error's selector to `expectRevert` but it didn't work

Comment (QGarchery) [reply test/MorphoTokenOptimism.t.sol:39]: It seems to work for me when I do `        vm.expectRevert(OptimismMintableERC20Upgradeable.ZeroAddress.selector);`

File: src/OptimismMintableERC20Upgradeable.sol
Comment (Jean-Grimal) [reply src/OptimismMintableERC20Upgradeable.sol:1]: I don't know how to write a test that implies two different networks

File: test/MorphoTokenOptimism.t.sol (hunk 6 above)
Comment (Jean-Grimal) [reply test/MorphoTokenOptimism.t.sol:39]: Oh cool
I'll make the change in another PR then (for all the tests)

File: src/ERC20DelegatesUpgradeable.sol (hunk 2 above)
Comment (MerlinEgalite) [reply src/ERC20DelegatesUpgradeable.sol:15]: Yes + we want to transition more and more to the Association in general

File: test/MorphoToken.t.sol (hunk 8 above)
Comment (Jean-Grimal) [reply test/MorphoToken.t.sol:208]: Not a strong opinion on this

File: src/OptimismMintableERC20Upgradeable.sol [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        return _interfaceId == iface1 || _interfaceId == iface3;
+    }
```
[the remark this answers is not in the document]
Comment (Jean-Grimal) [reply src/OptimismMintableERC20Upgradeable.sol]: I don't have a strong opinion on this

File: src/ERC20DelegatesUpgradeable.sol [hunk 10]
```diff
@@ -12,13 +12,13 @@ import {EIP712Upgradeable} from
...
 /// @custom:contact security@morpho.org
 /// @dev Extension of ERC20 to support token delegation.
 ///
 /// This extension keeps track of each account's vote power. Vote power can be delegated either by calling the
```
[the remark this answers is not in the document]
Comment (QGarchery) [reply src/ERC20DelegatesUpgradeable.sol:19]: I would truly love it but the current convention is to wrap lines in code comments (contrary to text files). Btw it means we should add the following to `foundry.toml` to enforce it:
```toml
[profile.default.fmt]
wrap_comments = true
```

File: src/MorphoTokenOptimism.sol [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    /// @dev The symbol of the token.
+    string internal constant SYMBOL = "MORPHO";
+
+    /* PUBLIC */
+
+    /// @notice Initializes the contract.
+    /// @param dao The DAO address.
+    /// @param remoteToken The address of the Morpho Token on Ethereum.
+    /// @param bridge The address of the StandardBridge contract.
+    function initialize(address dao, address remoteToken, address bridge) public initializer {
```
Comment (colin-morpho) [inline src/MorphoTokenOptimism.sol:36]: Shouldn't we call `_disableInitializers` in the constructor as well here?

File: src/OptimismMintableERC20Upgradeable.sol [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        emit Burn(from, amount);
+    }
+
+    /// @notice ERC165 interface check function.
+    /// @param _interfaceId Interface ID to check.
+    /// @return Whether or not the interface is supported by this contract.
+    function supportsInterface(bytes4 _interfaceId) external pure returns (bool) {
+        bytes4 interfaceERC165 = type(IERC165).interfaceId;
+        bytes4 interfaceOptimismMintableERC20 = type(IOptimismMintableERC20).interfaceId;
+        return _interfaceId == interfaceERC165 || _interfaceId == interfaceOptimismMintableERC20;
```
Comment (colin-morpho) [inline src/OptimismMintableERC20Upgradeable.sol:95]: Can't we merge this in a single mask?

File: src/MorphoTokenOptimism.sol (hunk 11 above)
Comment (colin-morpho) [reply src/MorphoTokenOptimism.sol:36]: In fact, can't we generate the Optimizer version from the other one?

File: src/OptimismMintableERC20Upgradeable.sol (hunk 12 above)
Comment (QGarchery) [reply src/OptimismMintableERC20Upgradeable.sol:95]: I assume you mean doing:
```solidity
bytes4 aggregatedInterface = interfaceERC165 ^ interfaceOptimismMintableERC20;
return _interfaceId == aggregatedInterface
```
Then no we can't do that because it wouldn't comply to ERC165, that specifies that `supportsInterface` has to return true on `interfaceERC165`

File: src/MorphoTokenOptimism.sol (hunk 11 above)
Comment (MerlinEgalite) [reply src/MorphoTokenOptimism.sol:36]: @Jean-Grimal can you make this happen?

File: src/MorphoTokenOptimism.sol [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        address _bridge;
+    }
+
+    /* EVENTS */
+
+    /// @dev Emitted whenever tokens are minted for an account.
+    event Mint(address indexed account, uint256 amount);
+
+    /// @dev Emitted whenever tokens are burned from an account.
+    event Burn(address indexed account, uint256 amount);
```
Comment (colin-morpho) [inline src/MorphoTokenOptimism.sol:41]: Move these to Token.sol.

File: src/ERC20DelegatesUpgradeable.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (QGarchery) [reply src/ERC20DelegatesUpgradeable.sol]: done in #45

File: test/MorphoToken.t.sol (hunk 8 above)
Comment (QGarchery) [reply test/MorphoToken.t.sol]: done in #45

File: src/Token.sol
Comment (QGarchery) [inline src/Token.sol:1]: It could make sense to have the `_authorizeUpgrade` function be virtual override, to make it easier to reuse

Comment (QGarchery) [inline src/Token.sol:1]: What about making the pragma of this contract more loose ? Like ^0.8.27, so that it could be reeused easier. Note that it wouldn't change the fact that the version is still constrained to 0.8.27 for contracts that are planned for deployment (`MorphoTokenEthereum`, `MorphoTokenOptimism`, and `Wrapper`)

File: src/MorphoTokenOptimism.sol (hunk 13 above)
Comment (QGarchery) [reply src/MorphoTokenOptimism.sol:41]: Handled on the PRs on top

**Review by QGarchery — APPROVED** [review]
Approving if both #42 and #43 are merged

File: src/Token.sol
Comment (MerlinEgalite) [reply src/Token.sol:1]: very good point

File: src/MorphoTokenOptimism.sol [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    // keccak256(abi.encode(uint256(keccak256("morpho.storage.OptimismMintableERC20")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 internal constant OptimismMintableERC20StorageLocation =
+        0x6fd4c0a11d0843c68c809f0a5f29b102d54bc08a251c384d9ad17600bfa05d00;
+
+    /* STORAGE LAYOUT */
+
+    /// @custom:storage-location erc7201:morpho.storage.OptimismMintableERC20
+    struct OptimismMintableERC20Storage {
+        address _remoteToken;
+        address _bridge;
```
Comment (MathisGD) [inline src/MorphoTokenOptimism.sol:32]: couldn't they both be immutable? like in https://docs.optimism.io/builders/app-developers/tutorials/standard-bridge-custom-token

the token is upgradable in case

File: src/MorphoTokenOptimism.sol [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    /* ERRORS */
+
+    /// @notice Reverts if the address is the zero address.
+    error ZeroAddress();
+
+    /// @notice Reverts if the caller is not the bridge.
+    error NotBridge(address caller);
```
Comment (MathisGD) [inline src/MorphoTokenOptimism.sol:41]:
```suggestion
    error NotBridge();
```

File: src/MorphoTokenOptimism.sol [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        address _remoteToken;
+        address _bridge;
+    }
+
+    /* ERRORS */
+
+    /// @notice Reverts if the address is the zero address.
+    error ZeroAddress();
+
+    /// @notice Reverts if the caller is not the bridge.
```
Comment (MathisGD) [inline src/MorphoTokenOptimism.sol:40]:
```suggestion
    /// @notice Thrown if the caller is not the bridge.
```

File: src/MorphoTokenOptimism.sol [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+    /// @custom:storage-location erc7201:morpho.storage.OptimismMintableERC20
+    struct OptimismMintableERC20Storage {
+        address _remoteToken;
+        address _bridge;
+    }
+
+    /* ERRORS */
+
+    /// @notice Reverts if the address is the zero address.
```
Comment (MathisGD) [inline src/MorphoTokenOptimism.sol:37]:
```suggestion
    /// @notice Thrown if the address is the zero address.
```

File: src/MorphoTokenOptimism.sol [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    /// @notice ERC165 interface check function.
+    /// @param _interfaceId Interface ID to check.
+    /// @return Whether or not the interface is supported by this contract.
+    function supportsInterface(bytes4 _interfaceId) external pure returns (bool) {
+        bytes4 interfaceERC165 = type(IERC165).interfaceId;
+        bytes4 interfaceOptimismMintableERC20 = type(IOptimismMintableERC20).interfaceId;
+        return _interfaceId == interfaceERC165 || _interfaceId == interfaceOptimismMintableERC20;
+    }
```
Comment (MathisGD) [inline src/MorphoTokenOptimism.sol:92]:
```suggestion
    function supportsInterface(bytes4 _interfaceId) external pure returns (bool) {
        return _interfaceId == type(IERC165).interfaceId || _interfaceId == type(IOptimismMintableERC20).interfaceId;
    }
```
